# Exploring the dataset of used cars from eBay Kleinanzeigen

# Table of Contens

- [Introduction](#intro)
- [Data Cleaning](#clean)
- [Initial Exploration and Cleaning](#explore)
- [Exploring the Odometer and Price Columns](#price)
- [Exploring the Date Columns](#date)
- [Dealing with Incorrect Registration Year Data](#reg)
- [Exploring Price by Brand](#brand)
- [Storing Aggregate Data in a DataFrame](#df)

# Introduction<a id="intro"></a>

We Will work on the dataSet of used cars from eBay Kleinanzeigen, German eBay website.
It was uploaded on [dataset](https://www.kaggle.com/datasets/sijovm/used-cars-data-from-ebay-kleinanzeigen?resource=download).

`Aims`
- To clean the dataset, perform initial analysis
- Find more on the way

In [49]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding='latin1')

In [50]:
autos.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [51]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

- We have imported `Pandas` and `Numpy` libraries
- Observations suggest, `50000 Entries` of dataset Is present to us
- The dataframe has `20` columns

## Data Cleaning <a id="clean"></a>

- We will use pythons preferred snaKecase instead of camelcase

In [52]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [53]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [54]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

- Some modifications done, and we changed into snake_case

In [55]:
autos.head(4)

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28

Now the titles are more redeable and python friendly, To note, we changed some Long names for our convinience:
- `yearOfRegistration` to `registration_year`
- `monThOfRegistration` to `registration_month`
- `notRepairedDamage` to `unrepaired_damage`
- `dateCreated` to `ad_created`
- `camelCase` to `snaKe_case` for the rest

## Initial Exploration and Cleaning<a id="explore"></a>

In [56]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-19 17:36:18  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Some columns have less counts than what was expected.
Like the vehicle_types has 44905 counts Instead 50000, MayBe some values are omitted here. Also, gearbox, model, fuel_type, unrepaired_damage also have the same status.
Columns, registration_year, power_ps, registration_month, nr_of_pictures, and postal_code, are showing NaN values.

In [57]:
autos.value_counts()

date_crawled         name                                                           seller  offer_type  price    abtest   vehicle_type  registration_year  gearbox    power_ps  model        odometer   registration_month  fuel_type  brand          unrepaired_damage  ad_created           nr_of_pictures  postal_code  last_seen          
2016-04-07 14:36:56  Mercedes_Benz_GLK_220_CDI_DPF_4Matic_BlueEFFICIENCY_7G_TRONIC  privat  Angebot     $25,900  control  suv           2009               automatik  170       glk          80,000km   7                   diesel     mercedes_benz  nein               2016-04-07 00:00:00  0               27299        2016-04-07 14:36:56    1
2016-03-05 14:06:30  VW_Multivan_T4_2.5_Automatik_/_auch_Behindertengerecht!        privat  Angebot     $3,950   test     bus           1995               automatik  110       transporter  150,000km  6                   benzin     volkswagen     nein               2016-03-05 00:00:00  0               65719        2016-03-13

The value for unrepaired_damage, Is showing nein, Which means no in German

In [58]:
autos['unrepaired_damage']

0        nein
1        nein
2        nein
3        nein
4        nein
         ... 
49995    nein
49996    nein
49997    nein
49998    nein
49999    nein
Name: unrepaired_damage, Length: 50000, dtype: object

price and odometer columns are Having numeric values, but, in sTring formats, so, We convert These columns into numeric values either Integer or floats

## Exploring the Odometer and Price Columns<a id="price"></a>

In [59]:
autos[['price', 'odometer']]

price   odometer
0       $5,000  150,000km
1       $8,500  150,000km
2       $8,990   70,000km
3       $4,350   70,000km
4       $1,350  150,000km
...        ...        ...
49995  $24,900  100,000km
49996   $1,980  150,000km
49997  $13,200    5,000km
49998  $22,900   40,000km
49999   $1,250  150,000km

[50000 rows x 2 columns]

In [60]:
autos['price'] = autos['price'].str.replace('$', '')
autos['price'] = autos['price'].str.replace(',', '')
autos['price'] = autos['price'].astype(float)
print(type(autos.iloc[4, 0]))

<class 'str'>


In [61]:
autos['odometer'] = autos['odometer'].str.replace('km', '')
autos['odometer'] = autos['odometer'].str.replace(',', '')
autos['odometer'] = autos['odometer'].astype(float)
autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)

In [62]:
type(autos['odometer_km'])

pandas.core.series.Series

In [63]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  5000.0  control          bus               2004   
1  privat    Angebot  8500.0  control    limousine               1997   
2  privat    Angebot  8990.0     test    limousine               2009   
3  privat    Angebot  4350.0  control   kleinwagen               2007   
4  privat    Angebot  1350.0     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere     150000.0                   3       lpg   
1  automatik       286     7er     150000.0                   6    benzin   
2    manuell       102    golf      70000.0                   7    benzin   
3  automatik        71  fortwo      70000.0                   6    benzin   
4    manuell         0   focus     150000.0                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [64]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   date_crawled        50000 non-null  object 
 1   name                50000 non-null  object 
 2   seller              50000 non-null  object 
 3   offer_type          50000 non-null  object 
 4   price               50000 non-null  float64
 5   abtest              50000 non-null  object 
 6   vehicle_type        44905 non-null  object 
 7   registration_year   50000 non-null  int64  
 8   gearbox             47320 non-null  object 
 9   power_ps            50000 non-null  int64  
 10  model               47242 non-null  object 
 11  odometer_km         50000 non-null  float64
 12  registration_month  50000 non-null  int64  
 13  fuel_type           45518 non-null  object 
 14  brand               50000 non-null  object 
 15  unrepaired_damage   40171 non-null  object 
 16  ad_c

In [65]:
autos.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                 50000        50000   50000      50000  5.000000e+04   
unique                48213        38754       2          2           NaN   
top     2016-03-19 17:36:18  Ford_Fiesta  privat    Angebot           NaN   
freq                      3           78   49999      49999           NaN   
mean                    NaN          NaN     NaN        NaN  9.840044e+03   
std                     NaN          NaN     NaN        NaN  4.811044e+05   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.100000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  1.000000e+08   

       abtest vehicle_type  registration_year  gearbox      power_ps  model  \
count   50000        44905       50000.000000    47320  50000.000000  47242   
unique      2            8                NaN        2           NaN    245   
top      test    limousine                NaN  manuell           NaN   golf   
freq    25756        12859                NaN    36993           NaN   4024   
mean      NaN          NaN        2005.073280      NaN    116.355920    NaN   
std       NaN          NaN         105.712813      NaN    209.216627    NaN   
min       NaN          NaN        1000.000000      NaN      0.000000    NaN   
25%       NaN          NaN        1999.000000      NaN     70.000000    NaN   
50%       NaN          NaN        2003.000000      NaN    105.000000    NaN   
75%       NaN          NaN        2008.000000      NaN    150.000000    NaN   
max       NaN          NaN        9999.000000      NaN  17700.000000    NaN   

          odometer_km  registration_month fuel_type       brand  \
count    50000.000000        50000.000000     45518       50000   
unique            NaN                 NaN         7          40   
top               NaN                 NaN    benzin  volkswagen   
freq              NaN                 NaN     30107       10687   
mean    125732.700000            5.723360       NaN         NaN   
std      40042.211706            3.711984       NaN         NaN   
min       5000.000000            0.000000       NaN         NaN   
25%     125000.000000            3.000000       NaN         NaN   
50%     150000.000000            6.000000       NaN         NaN   
75%     150000.000000            9.000000       NaN         NaN   
max     150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count              40171                50000         50000.0  50000.000000   
unique                 2                   76             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq               35232                 1946             NaN           NaN   
mean                 NaN                  NaN             0.0  50813.627300   
std                  NaN                  NaN             0.0  25779.747957   
min                  NaN                  NaN             0.0   1067.000000   
25%                  NaN                  NaN             0.0  30451.000000   
50%                  NaN                  NaN             0.0  49577.000000   
75%                  NaN                  NaN             0.0  71540.000000   
max                  NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

Still the correct results are not showing In `price` and `odometer_km` columns
- lets search for outliers

In [66]:
autos['price'].unique().shape

(2357,)

In [67]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [68]:
autos['price']

0         5000.0
1         8500.0
2         8990.0
3         4350.0
4         1350.0
          ...   
49995    24900.0
49996     1980.0
49997    13200.0
49998    22900.0
49999     1250.0
Name: price, Length: 50000, dtype: float64

In [69]:
autos['price'].value_counts().head()

price
0.0       1421
500.0      781
1500.0     734
2500.0     643
1000.0     639
Name: count, dtype: int64

In [70]:
autos['price'].sort_index()

0         5000.0
1         8500.0
2         8990.0
3         4350.0
4         1350.0
          ...   
49995    24900.0
49996     1980.0
49997    13200.0
49998    22900.0
49999     1250.0
Name: price, Length: 50000, dtype: float64

In [71]:
autos['price'].sort_values()

40874           0.0
15903           0.0
9812            0.0
20940           0.0
26392           0.0
            ...    
27371    12345678.0
47598    12345678.0
39377    12345678.0
42221    27322222.0
39705    99999999.0
Name: price, Length: 50000, dtype: float64

In [72]:
autos['price'].value_counts()

price
0.0        1421
500.0       781
1500.0      734
2500.0      643
1000.0      639
           ... 
16845.0       1
15470.0       1
18890.0       1
5198.0        1
1960.0        1
Name: count, Length: 2357, dtype: int64

In [73]:
autos['price'].sort_values(ascending=False).value_counts().sort_index(ascending=False).head(50)

price
99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
194000.0      1
190000.0      1
180000.0      1
175000.0      1
169999.0      1
169000.0      1
163991.0      1
163500.0      1
155000.0      1
151990.0      1
145000.0      1
139997.0      1
137999.0      1
135000.0      1
130000.0      1
129000.0      1
128000.0      1
120000.0      2
119900.0      1
119500.0      1
116000.0      1
115991.0      1
115000.0      1
114400.0      1
109999.0      1
105000.0      2
104900.0      1
99900.0       2
99000.0       2
98500.0       1
Name: count, dtype: int64

In [74]:
autos[autos['price']> 100000].value_counts().head()

date_crawled         name                                                 seller  offer_type  price       abtest   vehicle_type  registration_year  gearbox    power_ps  model   odometer_km  registration_month  fuel_type  brand          unrepaired_damage  ad_created           nr_of_pictures  postal_code  last_seen          
2016-03-05 17:39:22  Mercedes_Benz_SLS_AMG_Roadster_AMG_SPEEDSHIFT_DCT    privat  Angebot     180000.0    test     cabrio        2012               automatik  571       andere  50000.0      2                   benzin     mercedes_benz  nein               2016-03-05 00:00:00  0               14193        2016-03-20 19:44:36    1
2016-03-05 18:40:13  Porsche_911_991_GT3_LED_Licht_Chrono_Navi_Clubsport  privat  Angebot     163991.0    test     coupe         2014               automatik  476       911     10000.0      6                   benzin     porsche        nein               2016-03-05 00:00:00  0               51702        2016-03-09 16:45:30    1
2016-03-12 16:5

We will keep the Values at minimum 500, and Maximum up to 300000

In [75]:
autos = autos[autos['price'].between(500, 130000)]

In [76]:
autos[autos['price']>100000].head()

date_crawled                                               name  \
1878   2016-03-12 16:58:37                                  Porsche_911_Turbo   
2751   2016-03-15 10:52:35                               Porsche_911___993_4S   
7402   2016-03-22 19:48:09  Porsche_911_Carrera_4S_Cabrio_PDK__BOSE__NEU__...   
8232   2016-04-01 21:50:47               Porsche_993_S_Schalter_BRD_neuwertig   
16964  2016-04-01 16:46:18                    Bentley_Continental_Supersports   

       seller offer_type     price   abtest vehicle_type  registration_year  \
1878   privat    Angebot  129000.0  control        coupe               1995   
2751   privat    Angebot  120000.0  control        coupe               1998   
7402   privat    Angebot  115000.0     test       cabrio               2016   
8232   privat    Angebot  128000.0     test        coupe               1997   
16964  privat    Angebot  105000.0  control        coupe               2010   

         gearbox  power_ps model  odometer_km  registration_month fuel_type  \
1878     manuell       408   911     125000.0                   9    benzin   
2751     manuell       286   911     125000.0                   3    benzin   
7402   automatik       400   911       5000.0                   3    benzin   
8232     manuell       286   911     100000.0                   4    benzin   
16964  automatik       630   NaN      80000.0                   1    benzin   

                brand unrepaired_damage           ad_created  nr_of_pictures  \
1878          porsche              nein  2016-03-12 00:00:00               0   
2751          porsche              nein  2016-03-15 00:00:00               0   
7402          porsche              nein  2016-03-22 00:00:00               0   
8232          porsche              nein  2016-04-01 00:00:00               0   
16964  sonstige_autos              nein  2016-04-01 00:00:00               0   

       postal_code            last_seen  
1878         70180  2016-04-05 04:49:19  
2751         25488  2016-04-05 19:47:31  
7402         51379  2016-03-26 21:46:46  
8232         81543  2016-04-05 19:46:23  
16964        70192  2016-04-05 11:49:31

In [77]:
autos['price'].describe()

count     45073.000000
mean       6214.237126
std        7899.323832
min         500.000000
25%        1500.000000
50%        3500.000000
75%        7900.000000
max      130000.000000
Name: price, dtype: float64

Now the discriptive statistics looks more realistic to work upon

Odometer range can be taken From 0 to 322,000 km, but, others Are taken outliers

In [78]:
autos[autos['odometer_km']>322000].value_counts()

Series([], Name: count, dtype: int64)

From the observation, it can be concluded that no alternations are needeed in `odometer_km` column.

## Exploring the Date Columns <a id="date"></a>

We will work with date columns, `data_crawled`, `ad_created`, `last_seen`

In [79]:
autos[['date_crawled', 'ad_created', 'last_seen']]

date_crawled           ad_created            last_seen
0      2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1      2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2      2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3      2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4      2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50
...                    ...                  ...                  ...
49995  2016-03-27 14:38:19  2016-03-27 00:00:00  2016-04-01 13:47:40
49996  2016-03-28 10:50:25  2016-03-28 00:00:00  2016-04-02 14:18:02
49997  2016-04-02 14:44:48  2016-04-02 00:00:00  2016-04-04 11:47:27
49998  2016-03-08 19:25:42  2016-03-08 00:00:00  2016-04-05 16:45:07
49999  2016-03-14 00:42:12  2016-03-13 00:00:00  2016-04-06 21:18:48

[45073 rows x 3 columns]

In [80]:
autos.iloc[0]

date_crawled                       2016-03-26 17:47:46
name                  Peugeot_807_160_NAVTECH_ON_BOARD
seller                                          privat
offer_type                                     Angebot
price                                           5000.0
abtest                                         control
vehicle_type                                       bus
registration_year                                 2004
gearbox                                        manuell
power_ps                                           158
model                                           andere
odometer_km                                   150000.0
registration_month                                   3
fuel_type                                          lpg
brand                                          peugeot
unrepaired_damage                                 nein
ad_created                         2016-03-26 00:00:00
nr_of_pictures                                       0
postal_cod

In [86]:
autos['date_crawled']

0        2016-03-26 17:47:46
1        2016-04-04 13:38:56
2        2016-03-26 18:57:24
3        2016-03-12 16:58:10
4        2016-04-01 14:38:50
                ...         
49995    2016-03-27 14:38:19
49996    2016-03-28 10:50:25
49997    2016-04-02 14:44:48
49998    2016-03-08 19:25:42
49999    2016-03-14 00:42:12
Name: date_crawled, Length: 45073, dtype: object

In [89]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

date_crawled
2016-03-05    0.025536
2016-03-06    0.014133
2016-03-07    0.036208
2016-03-08    0.033168
2016-03-09    0.032924
2016-03-10    0.032725
2016-03-11    0.033035
2016-03-12    0.037339
2016-03-13    0.015530
2016-03-14    0.036319
2016-03-15    0.034034
2016-03-16    0.029352
2016-03-17    0.031149
2016-03-18    0.012868
2016-03-19    0.034766
2016-03-20    0.038072
2016-03-21    0.037717
2016-03-22    0.032991
2016-03-23    0.032392
2016-03-24    0.028975
2016-03-25    0.031083
2016-03-26    0.032658
2016-03-27    0.031149
2016-03-28    0.034810
2016-03-29    0.033279
2016-03-30    0.033302
2016-03-31    0.031660
2016-04-01    0.033901
2016-04-02    0.035786
2016-04-03    0.038826
2016-04-04    0.036629
2016-04-05    0.013156
2016-04-06    0.003173
2016-04-07    0.001353
Name: proportion, dtype: float64

For March and April 2016, the site seems to be visited almost everyday.

In [91]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

ad_created
2015-06-11    0.000022
2015-08-10    0.000022
2015-09-09    0.000022
2015-11-10    0.000022
2015-12-05    0.000022
                ...   
2016-04-03    0.039048
2016-04-04    0.037007
2016-04-05    0.011892
2016-04-06    0.003261
2016-04-07    0.001198
Name: proportion, Length: 76, dtype: float64

Ads are created With the intervals of some month, but, During April 2016, it can be seen, it was created on a daily basis. 

In [93]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

last_seen
2016-03-05    0.001087
2016-03-06    0.004171
2016-03-07    0.005214
2016-03-08    0.007011
2016-03-09    0.009451
2016-03-10    0.010294
2016-03-11    0.012047
2016-03-12    0.023917
2016-03-13    0.008874
2016-03-14    0.012291
2016-03-15    0.015686
2016-03-16    0.016152
2016-03-17    0.027688
2016-03-18    0.007388
2016-03-19    0.015419
2016-03-20    0.020411
2016-03-21    0.020655
2016-03-22    0.021254
2016-03-23    0.018415
2016-03-24    0.019546
2016-03-25    0.018592
2016-03-26    0.016462
2016-03-27    0.015442
2016-03-28    0.020522
2016-03-29    0.021365
2016-03-30    0.024161
2016-03-31    0.023451
2016-04-01    0.022874
2016-04-02    0.024871
2016-04-03    0.024937
2016-04-04    0.024316
2016-04-05    0.126572
2016-04-06    0.225346
2016-04-07    0.134116
Name: proportion, dtype: float64

Everyday some old ads are removed.

In [88]:
autos['registration_year'].describe()

count    45073.000000
mean      2005.062011
std         89.675606
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

2005 is the average year to have most cars registered.
Minimum year is seen to be 1000 and maximum year is 9999 that seems to be less realistic.

## Dealing with Incorrect Registration Year Data<a id="reg"></a>

We saw that 1000 and 9999 are some odd values that are outliers. The list cannot go less than 1900 and more than 2016.

In [96]:
autos['registration_year'].value_counts().sort_index()

registration_year
1000    1
1001    1
1910    2
1927    1
1929    1
       ..
5911    1
6200    1
8888    1
9000    1
9999    3
Name: count, Length: 93, dtype: int64

Removing those outliers dates

In [98]:
autos = autos[autos['registration_year'].between(1900, 2016)]

In [100]:
autos['registration_year'].describe()

count    43299.000000
mean      2003.224001
std          7.072819
min       1910.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

Now it looks more realistic.

In [104]:
autos['registration_year'].value_counts(normalize=True).sort_index().tail(10)

registration_year
2007    0.052403
2008    0.050925
2009    0.047992
2010    0.036560
2011    0.037345
2012    0.030116
2013    0.018361
2014    0.015035
2015    0.008314
2016    0.021848
Name: proportion, dtype: float64

2007 and 2008 has high Percentage, 5% For Registrations, yet, in 2016, it has decreased to 2%. Probably, New modals Are built less in numbers with the passage of time.

## Exploring Price by Brand<a id="brand"></a>

In [106]:
autos['brand']

0           peugeot
1               bmw
2        volkswagen
3             smart
4              ford
            ...    
49995          audi
49996          opel
49997          fiat
49998          audi
49999          opel
Name: brand, Length: 43299, dtype: object

In [107]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'renault', 'audi', 'sonstige_autos', 'mazda', 'porsche', 'mini',
       'mercedes_benz', 'seat', 'toyota', 'opel', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'jaguar', 'fiat', 'skoda', 'subaru',
       'kia', 'citroen', 'mitsubishi', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [119]:
autos['brand'].value_counts(normalize=True).head(100)

brand
volkswagen        0.212014
bmw               0.115014
mercedes_benz     0.102266
opel              0.100395
audi              0.090533
ford              0.065198
renault           0.043927
peugeot           0.029377
fiat              0.023534
seat              0.017899
skoda             0.017160
smart             0.015127
mazda             0.015012
nissan            0.015012
citroen           0.014157
toyota            0.013418
hyundai           0.010162
sonstige_autos    0.009654
volvo             0.009377
mini              0.009377
honda             0.007806
mitsubishi        0.007737
kia               0.007344
alfa_romeo        0.006628
porsche           0.006051
chevrolet         0.006028
suzuki            0.005912
chrysler          0.003603
dacia             0.002841
jeep              0.002448
land_rover        0.002263
daihatsu          0.002171
subaru            0.001986
saab              0.001663
jaguar            0.001617
daewoo            0.001340
rover             0.00

Lets only take Top brands which has more than 5% of preference of the marketplace.

Volkswagen(21%), bmw(12%), mercedez_benz(10%) are the preferred brands of all the time, while, trabant, lancia, lada are the least preferred brands of cars. 

In [121]:
brand_proportion = autos['brand'].value_counts(normalize=True)
brand_proportion

brand
volkswagen        0.212014
bmw               0.115014
mercedes_benz     0.102266
opel              0.100395
audi              0.090533
ford              0.065198
renault           0.043927
peugeot           0.029377
fiat              0.023534
seat              0.017899
skoda             0.017160
smart             0.015127
mazda             0.015012
nissan            0.015012
citroen           0.014157
toyota            0.013418
hyundai           0.010162
sonstige_autos    0.009654
volvo             0.009377
mini              0.009377
honda             0.007806
mitsubishi        0.007737
kia               0.007344
alfa_romeo        0.006628
porsche           0.006051
chevrolet         0.006028
suzuki            0.005912
chrysler          0.003603
dacia             0.002841
jeep              0.002448
land_rover        0.002263
daihatsu          0.002171
subaru            0.001986
saab              0.001663
jaguar            0.001617
daewoo            0.001340
rover             0.00

In [130]:
most_preferred_brands = brand_proportion[brand_proportion > 0.05].index
# We only need index
most_preferred_brands

Index(['volkswagen', 'bmw', 'mercedes_benz', 'opel', 'audi', 'ford'], dtype='object', name='brand')

Lets get average price for these brands.
- Create a dictionary with brand name as key and Average price as the value, for key_value pairs

In [143]:
best_cars_prices = {}

for brand in most_preferred_brands:
    the_brand = autos[autos['brand'] == brand]
#     This is to copy aggregate of The brand data as a whole all the rows of volswagon Brand for example
    mean_price = the_brand['price'].mean()
    
    best_cars_prices[brand] = float(mean_price)

In [144]:
best_cars_prices

{'volkswagen': 5783.622984749455,
 'bmw': 8447.069879518072,
 'mercedes_benz': 8691.820234869016,
 'opel': 3394.0395675178283,
 'audi': 9571.457397959184,
 'ford': 4291.666312433581}

In [146]:
autos['price'].describe()

count     43299.000000
mean       6309.083074
std        7982.928717
min         500.000000
25%        1500.000000
50%        3500.000000
75%        7999.000000
max      130000.000000
Name: price, dtype: float64

`volkswagen` with average price 5784 that is Preferred more. It is less expesve than less preferred ones. Even, the high priced cars are less preferrd. In average car's price starts from 6309, but, the Avg_price for the most preferrd onE, `volkwagen` is even less than The total average price of Cars. 

## Storing Aggregate Data in a DataFrame<a id="df"></a>

Now Lets understand The Link between mileage and price of The car.

In [172]:
bcp_series = pd.Series(best_cars_prices)
bcp_df = pd.DataFrame(bcp_series, columns=['mean_price'])

In [173]:
bcp_df

mean_price
volkswagen     5783.622985
bmw            8447.069880
mercedes_benz  8691.820235
opel           3394.039568
audi           9571.457398
ford           4291.666312

We just created a Pandas DataFrame from the previous data.

In [158]:
top_mean_mileage = {}

# Lets Use a Loop and Create aggregate To repreSent all the simiLar rows based on The same brand
for brand in most_preferred_brands:
    the_brand = autos[autos['brand'] == brand]
#     Aggregate brand
    
    mean_mileage_for_the_current_brand = the_brand['odometer_km'].mean()
    
    top_mean_mileage[brand] = int(mean_mileage_for_the_current_brand)

In [159]:
top_mean_mileage

{'volkswagen': 128234,
 'bmw': 132928,
 'mercedes_benz': 131058,
 'opel': 128012,
 'audi': 128941,
 'ford': 123494}

`bmw`And `mercedes_benz` Gives the more Mileage than any other. 
Other top brands are Seen to give good mileages too.

Convert The Dictionary into DataFrame

In [171]:
mean_mileage_series = pd.Series(top_mean_mileage)
mean_mileage_df = pd.DataFrame(mean_mileage_series, columns=['mean_mileage'])
mean_mileage_df

mean_mileage
volkswagen           128234
bmw                  132928
mercedes_benz        131058
opel                 128012
audi                 128941
ford                 123494

Lets combine both These infos, but, we will use The bcp_df Dataframe, and add mean_Mileage_series to the dateframe.

In [174]:
bcp_df['mean_mileage']= mean_mileage_series
# Like a dictionary addition
bcp_df

mean_price  mean_mileage
volkswagen     5783.622985        128234
bmw            8447.069880        132928
mercedes_benz  8691.820235        131058
opel           3394.039568        128012
audi           9571.457398        128941
ford           4291.666312        123494

The motive was to keep every Thing in one place for a proper analysis of The dataframe. We can now Visualize More Easily to see The same info that we described aboveAlready. The conClusion is Still, bmw Being second most preferred car with The all the time highest average Mileage. If you Want more mileage, bmw is a suitable one for you. 
Should you go For the publics choice, then, volkswagen is most preferred one, mayBe, it gives more qualIty and comfort, asn also, the Average price is at its best, not too low, or high, to Consider amonG top 6. 
Opel being the least expensive in average, gives pretty good mileage, and you can go for it, if you are tight on budget. Likewise, Ford is next In the line for the upGradable range of price. Overall, the Average price Is highest for audi, that is the most expeNsive car of all, and it comes 4th in the line for peOples's choice.